lib/extend.es6

```diff
@@ -31,38 +30,33 @@ export default (options = {}) => {
     };
 };
 
-const api = new Api;
-let baseFileOpts;
-
-function handleExtendsNodes(tree, options) {
-    baseFileOpts = baseFileOpts || tree.options;
-    tree.options = tree.options || baseFileOpts;
-
-    if (!tree.messages) {
-        tree.messages = [];
-    }
+const api = new Api();
 
+function handleExtendsNodes(tree, options, messages) {
     api.match.call(applyPluginsToTree(tree, options.plugins), {tag: 'extends'}, extendsNode => {
         if (! extendsNode.attrs || ! extendsNode.attrs.src) {
             throw getError(errors.EXTENDS_NO_SRC);
         }
 
         const layoutPath = path.resolve(options.root, extendsNode.attrs.src);
         const layoutHtml = fs.readFileSync(layoutPath, options.encoding);
-        let layoutTree;
+        let layoutTree;        
 
-        tree.messages.push({
+        layoutTree = handleExtendsNodes(applyPluginsToTree(parseToPostHtml(layoutHtml), options.plugins), options, messages);
+        extendsNode.tag = false;
+        extendsNode.content = mergeExtendsAndLayout(layoutTree, extendsNode);
+
+        messages.push({
             type: 'dependency',
             file: layoutPath,
-            from: tree.options.from
+            from: options.from
```

Comment:
This is another thing I don't quite understand:  ```options``` here is plugin options, but ```from``` property usually resides in ```tree.options```. I.e. in case of recursive extend ```from``` property should be the same in all extend "levels". So saving "base" file/tree options with its ```from``` property makes sense to me. But I probably wrong because I am not understanding the whole ecosystem fully. Here is related discussion: https://github.com/posthtml/posthtml/issues/260

Reply:
```
|– src
| |– extends
| | | – layout.html
| |– 2.html
| |– 1.html
```

```js
[ 'src/1.html', 'src/2.html' ].map((file) => {
 return posthtml([ extend(options) ]).process(file.content, { from: file.path })
})
```

#### `First Run/File`

```
tree.options = { from: 'src/1.html' }
tree.messages = [ { type: 'dependency', file: 'layoutPath', from: 'src/1.html' } ]
```
#### `Second Run/File`

```
tree.options = { from: 'src/2.html' }
tree.messages = [ { type: 'dependency', file: 'layoutPath', from: 'src/2.html' } ]
```

> ℹ️  `posthtml-loader` doesn't care about `msg.from`, only `msg.type` && `msg.file` are important (the imported layout which needs to be added to the watcher)

Reply:
Yeah, it makes sense. What I meant is that in case of recursive extend (```index.html``` extends ```layout2.html``` extends ```layout1.html```) there is one "initial file" ("initial ```tree```" which gets processed by posthtml initially) — ```index.html```, so it seemed logical to have ```layout1Subtree.options.from === layout2Subtree.options.from === indexTree.options.from```, because subtree is eventually inserted into the tree of "initial file". That's why I saved initial ```tree.options``` to set them into subtrees, probably mistakenly
```
|– src
| |– extends
| | | – layout1.html
| | | – layout2.html
| |– index.html
```
But probably it all doesn't matter considering that ```from``` is actually not important... (nevertheless ```from``` was/is still used here and in ```posthtml-include```)

Reply:
You mean 

```
|– src
| |– extends
| | => | – layout1.html (2)
| | | =>| – layout2.html (3)
| |– index.html (1)
```

?

That shouldn't be much of a problem since `dependencies` should be flatten via/by `result.messages` and the hierarchy is expressed by the plugin logic, which is still in place if the watcher triggers a rebuild when a dependency changes. Every time a plugin loads a subtree simply use a `posthtml` instance to parse and process and `concat` the `result.messages` of the importee with `tree.messages` of the importer.